File: MorkosiaPrepaLeague/Models/Core/LeagueOwnershipState.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace MorkosiaPrepaLeague.Models.Core
{
    /// <summary>
    /// Aggregator document that tracks player ownership counts for a specific transfer window.
    /// Used for optimistic concurrency control when validating team selections.
    /// </summary>
    public class LeagueOwnershipState : CosmosDocumentBase
    {
        [Required]
        public string LeagueId { get; set; } = string.Empty;
        
        [Required]
        public string TransferWindowId { get; set; } = string.Empty;
        
        /// <summary>
        /// Dictionary tracking ownership counts: playerId -> position -> count
        /// Example: { "player1": { "GK": 2, "DEF": 1 }, "player2": { "ATT": 3 } }
        /// </summary>
        public Dictionary<string, Dictionary<string, int>> OwnershipCounts { get; set; } = new();
        
        /// <summary>
        /// Total number of teams that have been submitted for this transfer window
        /// </summary>
        public int TotalTeamsCount { get; set; } = 0;
        
        /// <summary>
        /// Timestamp when this ownership state was last updated
        /// </summary>
        public DateTime LastUpdated { get; set; } = DateTime.UtcNow;
    }
}

File: src/app/models/league_ownership_state.py
from datetime import datetime, timezone
from typing import Dict

from pydantic import Field

from app.models.base import DocumentBase


class LeagueOwnershipState(DocumentBase):
    
    """
    Stored in the `teams` collection (alongside Team documents, same as in .NET).
    Mirrors LeagueOwnershipState.cs.

    Tracks how many teams each player is selected on per position,
    for a specific transfer window. Used for optimistic concurrency control
    when managers submit their team picks simultaneously.

    `id` format: "ownership-{transfer_window_id}"
    `ownership_counts` format: { player_id: { position: count } }
    """

    league_id: str = ""
    transfer_window_id: str = ""
    ownership_counts: Dict[str, Dict[str, int]] = {}
    total_teams_count: int = 0
    last_updated: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "teams"  # Lives in the same collection as Team documents
